Sources/PostgresNIO/Connection/PostgresConnection+Configuration.swift

```diff
@@ -0,0 +1,280 @@
+        ///
+        /// This is provided for calling code which wants to manage the underlying connection transport on its
+        /// own, such as when tunneling a connection through SSH.
+        ///
+        /// - Parameters:
+        ///   - channel: The `NIOCore/Channel` to use. The channel must already be active and connected to an
+        ///     endpoint (i.e. `NIOCore/Channel/isActive` must be `true`).
+        ///   - tls: The TLS mode to use. Defaults to ``TLS-swift.struct/disable``.
+        public init(establishedChannel channel: Channel, username: String, password: String?, database: String?) {
+            self.init(endpointInfo: .configureChannel(channel), tls: .disable, username: username, password: password, database: database)
+        }
+
+        // MARK: - Implementation details
+        // TODO: Make these `private` once the deprecated ``Connection`` struct is gone.
+
+        /*private*/ enum EndpointInfo {
+            case configureChannel(Channel)
+            case bindUnixDomainSocket(path: String)
+            case connectTCP(host: String, port: Int)
+        }
+
+        /*private*/ var endpointInfo: EndpointInfo
+
+        /*private*/ init(endpointInfo: EndpointInfo, tls: TLS, username: String, password: String?, database: String?) {
+            self.endpointInfo = endpointInfo
+            self.tls = tls
+            self.username = username
+            self.password = password
+            self.database = database
+        }
+    }
+}
+
+// MARK: - Internal config details
+
+extension PostgresConnection {
+    /// A configuration object to bring the new ``PostgresConnection.Configuration`` together with
+    /// the deprecated configuration.
+    ///
+    /// TODO: Drop with next major release
+    struct InternalConfiguration {
+        enum Connection {
+            case unresolvedTCP(host: String, port: Int)
+            case unresolvedUDS(path: String)
+            case resolved(address: SocketAddress)
+            case bootstrapped(channel: Channel)
+        }
+
+        let connection: InternalConfiguration.Connection
+        let username: String?
+        let password: String?
+        let database: String?
+        let tls: Configuration.TLS
+        let options: Configuration.Options
+    }
+}
+
+extension PostgresConnection.InternalConfiguration {
+    init(_ config: PostgresConnection.Configuration) {
+        switch config.endpointInfo {
+        case .connectTCP(let host, let port): self.connection = .unresolvedTCP(host: host, port: port)
+        case .bindUnixDomainSocket(let path): self.connection = .unresolvedUDS(path: path)
+        case .configureChannel(let channel): self.connection = .bootstrapped(channel: channel)
+        }
+        self.username = config.username
+        self.password = config.password
+        self.database = config.database
+        self.tls = config.tls
+        self.options = config.options
+    }
+    
+    var serverNameForSNI: String? {
```

Comment:
can we rename this to `serverNameForTLS`? Since it also used for certificate validation?